collector/json_collector.go

```diff
@@ -0,0 +1,207 @@
+// for followup processing.
+func NewJSONCollector(redfishClient *gofish.APIClient, logger *slog.Logger, config *config.JSONCollectorConfig) (*JSONCollector, error) {
+	query, err := gojq.Parse(config.JQFilter)
+	if err != nil {
+		return nil, fmt.Errorf("jq parse error in collector creation: %w", err)
+	}
+	rawClient, err := redfishClient.Get(config.RedfishRoot)
+	if err != nil {
+		return nil, fmt.Errorf("json collector could not perform lookup against %s: %w", config.RedfishRoot, err)
+	}
+	body, err := io.ReadAll(rawClient.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to cache json response: %w", err)
+	}
+
+	return &JSONCollector{
+		redfishClient:  redfishClient,
+		config:         config,
+		jqQuery:        query,
+		logger:         logger,
+		cachedResponse: body,
+		metricsDescs:   map[string]DescWithLabels{},
+	}, nil
+}
+
+// Describe implements prometheus.Collector
+func (j *JSONCollector) Describe(ch chan<- *prometheus.Desc) {
+	metrics, err := metricsFromBody(j.jqQuery, j.cachedResponse)
+	if err != nil {
+		j.logger.Error("failed to convert collected data to a metrics description", slog.Any("error", err))
+		return
+	}
+	for _, metric := range metrics {
+		labelNames := maps.Keys(metric.Labels)
+		sorted := slices.Sorted(labelNames)
+		d := prometheus.NewDesc(metric.Name, metric.Help, sorted, nil)
+		j.metricsDescs[metric.Name] = DescWithLabels{
+			SortedLabels: sorted,
+			Desc:         d,
+		}
+		ch <- d
+	}
+}
+
+// Collect implements prometheus.Collector
+func (j *JSONCollector) Collect(ch chan<- prometheus.Metric) {
+	metrics, err := metricsFromBody(j.jqQuery, j.cachedResponse)
+	if err != nil {
+		j.logger.Error("failed to convert collected data to metrics", slog.Any("error", err))
+		return
+	}
+	for _, metric := range metrics {
+		labels := []string{}
+		for _, desiredLabel := range j.metricsDescs[metric.Name].SortedLabels {
+			if lVal, ok := metric.Labels[desiredLabel]; ok {
+				labels = append(labels, lVal)
+			} else {
+				labels = append(labels, "")
+			}
+		}
+		ch <- prometheus.MustNewConstMetric(
+			j.metricsDescs[metric.Name].Desc,
+			prometheus.GaugeValue,
+			metric.Value,
+			labels...,
+		)
+	}
+}
+
+// metricsFromBody applies the given gojq.Query to a Redfish response body.
+// It is expected that the result of JQ application yields data in a format which may further be
+// converted to a typed struct. A slice of JSONYieldedMetric is returned then for all data which meets
+// this expectation.
+// An error during JQ parsing results skips the item, and errors encountered in this way
+// are joined together and returned as a bundle.
+func metricsFromBody(query *gojq.Query, jsonBody []byte) ([]JSONYieldedMetric, error) {
+	var yielded []JSONYieldedMetric
+	var parseErrors []error
+	var intermediary map[string]any
+
+	if err := json.Unmarshal(jsonBody, &intermediary); err != nil {
+		return yielded, err
+	}
+	iter := query.Run(intermediary)
```

Comment:
If we want to cancel an operation when a context is canceled in the middle of traversing a query, use `query.RunWithContext`.